File: Emulator/src/main/core/game/ge/GuidePrices.java
package core.game.ge;

import core.game.component.Component;
import core.game.node.entity.player.Player;
import core.tools.StringUtils;
import org.rs.consts.Components;

/**
 * The type Guide prices.
 */
public final class GuidePrices {

    private static final Component COMPONENT = new Component(Components.EXCHANGE_GUIDE_PRICE_642);

    /**
     * Open.
     *
     * @param player the player
     * @param type   the type
     */
    public static final void open(final Player player, final GuideType type) {
        player.getInterfaceManager().open(COMPONENT);
        type.display(player);
    }

    /**
     * Clear.
     *
     * @param player the player
     */
    public static final void clear(final Player player) {
        for (int i = 135; i < 165; i++) {
            player.getPacketDispatch().sendInterfaceConfig(COMPONENT.id, i, true);
        }
    }

    /**
     * The type Guide item.
     */
    public static class GuideItem {

        private final int item;
        private final int[] childData;

        /**
         * Instantiates a new Guide item.
         *
         * @param item      the item
         * @param childData the child data
         */
        public GuideItem(final int item, final int... childData) {
            this.item = item;
            this.childData = childData;
        }

        /**
         * Gets item.
         *
         * @return the item
         */
        public int getItem() {
            return item;
        }

        /**
         * Get child data int [ ].
         *
         * @return the int [ ]
         */
        public int[] getChildData() {
            return childData;
        }
    }

    /**
     * The enum Guide type.
     */
    public enum GuideType {

        /**
         * The Logs.
         */
        LOGS(new int[]{0, 0}, new GuideItem(1511, 155), new GuideItem(2862, 156), new GuideItem(1521, 157), new GuideItem(1519, 158),
            new GuideItem(6333, 159), new GuideItem(1517, 160), new GuideItem(6332, 161), new GuideItem(12581, 162),
            new GuideItem(1515, 163), new GuideItem(1513, 164)),

        /**
         * The Ores.
         */
        ORES(new int[]{40, 44}, new GuideItem(436, 33), new GuideItem(438, 34), new GuideItem(440, 35), new GuideItem(442, 36),
            new GuideItem(453, 37), new GuideItem(444, 38), new GuideItem(447, 39), new GuideItem(449, 40), new GuideItem(451, 41)),

        /**
         * The Runes.
         */
        RUNES(new int[]{215, 216}, new GuideItem(1436, 183), new GuideItem(7936, 184), new GuideItem(556, 185), new GuideItem(558, 186),
            new GuideItem(555, 187), new GuideItem(557, 188), new GuideItem(554, 189), new GuideItem(559, 190), new GuideItem(564, 191),
            new GuideItem(562, 192), new GuideItem(9075, 193), new GuideItem(561, 194), new GuideItem(563, 195), new GuideItem(560, 196),
            new GuideItem(565, 197), new GuideItem(566, 198)),

        /**
         * The Herbs.
         */
        HERBS(new int[]{130, 135}, new GuideItem(249, 119), new GuideItem(251, 120), new GuideItem(253, 121), new GuideItem(255, 122),
            new GuideItem(257, 123), new GuideItem(2998, 124), new GuideItem(259, 125), new GuideItem(12172, 126), new GuideItem(261, 127),
            new GuideItem(263, 128), new GuideItem(3000, 129), new GuideItem(265, 130), new GuideItem(2481, 131), new GuideItem(267, 132),
            new GuideItem(269, 133)),

        /**
         * The Weapons and armour.
         */
        WEAPONS_AND_ARMOUR(new int[]{88, 89}, new GuideItem(11834, 73), new GuideItem(11838, 74), new GuideItem(11842, 75),
            new GuideItem(11864, 76), new GuideItem(11870, 77), new GuideItem(11846, 78), new GuideItem(11848, 79),
            new GuideItem(11850, 80), new GuideItem(11856, 81), new GuideItem(11732, 82), new GuideItem(4151, 83),
            new GuideItem(11235, 84), new GuideItem(6739, 85), new GuideItem(4587, 86), new GuideItem(4153, 87));

        private final GuideItem[] items;
        private final int[] childData;

        GuideType(final int[] childData, final GuideItem... items) {
            this.childData = childData;
            this.items = items;
        }

        /**
         * Get items guide item [ ].
         *
         * @return the guide item [ ]
         */
        public GuideItem[] getItems() {
            return items;
        }

        /**
         * Get child data int [ ].
         *
         * @return the int [ ]
         */
        public int[] getChildData() {
            return childData;
        }

        /**
         * Display.
         *
         * @param player the player
         */
        public void display(final Player player) {
            player.setAttribute("guide-price", this);
            if (this != LOGS) {
                clear(player);
            }
            player.getPacketDispatch().sendString("Guide Prices: " + StringUtils.formatDisplayName(name()), COMPONENT.id, 14);
            for (int i = getChildData()[0]; i < getChildData()[1]; i++) {
                player.getPacketDispatch().sendInterfaceConfig(642, i, false);
            }
            for (GuideItem item : getItems()) {
                player.getPacketDispatch().sendString("" + GrandExchange.getRecommendedPrice(item.item, false) + " gp", COMPONENT.id, item.getChildData()[0]);
            }
        }
    }
}

File: Emulator/src/main/core/game/ge/GuidePrices.kt
package core.game.ge

import core.game.component.Component
import core.game.ge.GrandExchange.Companion.getRecommendedPrice
import core.game.node.entity.player.Player
import core.tools.StringUtils.formatDisplayName
import org.rs.consts.Components
import org.rs.consts.Items

/**
 * Provides functionality for managing and displaying guide prices in the Grand Exchange.
 */
object GuidePrices {
    private val COMPONENT = Component(Components.EXCHANGE_GUIDE_PRICE_642)

    /**
     * Opens the guide price interface for the specified player and type.
     *
     * @param player the player requesting the guide prices.
     * @param type   the type of guide prices to display.
     */
    @JvmStatic
    fun open(player: Player, type: GuideType) {
        player.interfaceManager.open(COMPONENT)
        type.display(player)
    }

    /**
     * Clears the guide price interface for the specified player.
     *
     * @param player the player whose guide price interface should be cleared.
     */
    fun clear(player: Player) {
        for (i in 135..164) {
            player.packetDispatch.sendInterfaceConfig(COMPONENT.id, i, true)
        }
    }

    /**
     * Represents an item in the guide price list.
     *
     * @param item      the item ID.
     * @param childData the child data for the item's display.
     */
    class GuideItem(
        /**
         * The item ID.
         */
        @JvmField val item: Int,

        /**
         * The child data for the item's display.
         */
        vararg val childData: Int
    )

    /**
     * Represents different types of guide price categories, such as logs, ores, etc.
     *
     * @param childData the range of child data indices to display for the guide type.
     * @param items     the items associated with the guide type.
     */
    enum class GuideType(
        /**
         * The child data indices for the guide type.
         */
        val childData: IntArray, vararg items: GuideItem
    ) {
        /**
         * Guide prices for logs.
         */
        LOGS(
            intArrayOf(0, 0),
            GuideItem(Items.LOGS_1511, 155),
            GuideItem(Items.ACHEY_TREE_LOGS_2862, 156),
            GuideItem(Items.OAK_LOGS_1521, 157),
            GuideItem(Items.WILLOW_LOGS_1519, 158),
            GuideItem(Items.TEAK_LOGS_6333, 159),
            GuideItem(Items.MAPLE_LOGS_1517, 160),
            GuideItem(Items.MAHOGANY_LOGS_6332, 161),
            GuideItem(Items.EUCALYPTUS_LOGS_12581, 162),
            GuideItem(Items.YEW_LOGS_1515, 163),
            GuideItem(Items.MAGIC_LOGS_1513, 164)
        ),

        /**
         * Guide prices for ores.
         */
        ORES(
            intArrayOf(40, 44),
            GuideItem(Items.COPPER_ORE_436, 33),
            GuideItem(Items.TIN_ORE_438, 34),
            GuideItem(Items.IRON_ORE_440, 35),
            GuideItem(Items.SILVER_ORE_442, 36),
            GuideItem(Items.COAL_453, 37),
            GuideItem(Items.GOLD_ORE_444, 38),
            GuideItem(Items.MITHRIL_ORE_447, 39),
            GuideItem(Items.ADAMANTITE_ORE_449, 40),
            GuideItem(Items.RUNITE_ORE_451, 41)
        ),

        /**
         * Guide prices for runes.
         */
        RUNES(
            intArrayOf(215, 216),
            GuideItem(Items.RUNE_ESSENCE_1436, 183),
            GuideItem(Items.PURE_ESSENCE_7936, 184),
            GuideItem(Items.AIR_RUNE_556, 185),
            GuideItem(Items.MIND_RUNE_558, 186),
            GuideItem(Items.WATER_RUNE_555, 187),
            GuideItem(Items.EARTH_RUNE_557, 188),
            GuideItem(Items.FIRE_RUNE_554, 189),
            GuideItem(Items.BODY_RUNE_559, 190),
            GuideItem(Items.COSMIC_RUNE_564, 191),
            GuideItem(Items.CHAOS_RUNE_562, 192),
            GuideItem(Items.ASTRAL_RUNE_9075, 193),
            GuideItem(Items.NATURE_RUNE_561, 194),
            GuideItem(Items.LAW_RUNE_563, 195),
            GuideItem(Items.DEATH_RUNE_560, 196),
            GuideItem(Items.BLOOD_RUNE_565, 197),
            GuideItem(Items.SOUL_RUNE_566, 198)
        ),

        /**
         * Guide prices for herbs.
         */
        HERBS(
            intArrayOf(130, 135),
            GuideItem(Items.CLEAN_GUAM_249, 119),
            GuideItem(Items.CLEAN_MARRENTILL_251, 120),
            GuideItem(Items.CLEAN_TARROMIN_253, 121),
            GuideItem(Items.CLEAN_HARRALANDER_255, 122),
            GuideItem(Items.CLEAN_RANARR_257, 123),
            GuideItem(Items.CLEAN_TOADFLAX_2998, 124),
            GuideItem(Items.CLEAN_IRIT_259, 125),
            GuideItem(Items.CLEAN_SPIRIT_WEED_12172, 126),
            GuideItem(Items.CLEAN_AVANTOE_261, 127),
            GuideItem(Items.CLEAN_KWUARM_263, 128),
            GuideItem(Items.CLEAN_SNAPDRAGON_3000, 129),
            GuideItem(Items.CLEAN_CADANTINE_265, 130),
            GuideItem(Items.CLEAN_LANTADYME_2481, 131),
            GuideItem(Items.CLEAN_DWARF_WEED_267, 132),
            GuideItem(Items.CLEAN_TORSTOL_269, 133)
        ),

        /**
         * Guide prices for weapons and armour.
         */
        WEAPONS_AND_ARMOUR(
            intArrayOf(88, 89),
            GuideItem(Items.ADAMANT_ARMOUR_SET_L_11834, 73),
            GuideItem(Items.RUNE_ARMOUR_SET_L_11838, 74),
            GuideItem(Items.DRAGON_CHAIN_ARMOUR_SET_L_11842, 75),
            GuideItem(Items.BLUE_DRAGONHIDE_SET_11866, 76),
            GuideItem(Items.BLACK_DRAGONHIDE_SET_11870, 77),
            GuideItem(Items.AHRIMS_SET_11846, 78),
            GuideItem(Items.DHAROKS_SET_11848, 79),
            GuideItem(Items.GUTHANS_SET_11850, 80),
            GuideItem(Items.VERACS_SET_11856, 81),
            GuideItem(Items.DRAGON_BOOTS_11732, 82),
            GuideItem(Items.ABYSSAL_WHIP_4151, 83),
            GuideItem(Items.DARK_BOW_11235, 84),
            GuideItem(Items.DRAGON_AXE_6739, 85),
            GuideItem(Items.DRAGON_SCIMITAR_4587, 86),
            GuideItem(Items.GRANITE_MAUL_4153, 87)
        );

        /**
         * The list of items for this guide type.
         */
        @JvmField
        val items: Array<GuideItem> = items as Array<GuideItem>

        /**
         * Displays the guide prices for the player.
         *
         * @param player the player to display the guide prices for.
         */
        fun display(player: Player) {
            player.setAttribute("guide-price", this)
            if (this != LOGS) {
                clear(player)
            }
            player.packetDispatch.sendString("Guide Prices: " + formatDisplayName(name), COMPONENT.id, 14)
            for (i in childData[0] until childData[1]) {
                player.packetDispatch.sendInterfaceConfig(Components.EXCHANGE_GUIDE_PRICE_642, i, false)
            }
            for (item in items) {
                player.packetDispatch.sendString(
                    "" + getRecommendedPrice(item.item, false) + " gp", COMPONENT.id, item.childData[0]
                )
            }
        }
    }
}
